### ISSUE TITLE
randomly stopped working on OSX

### ISSUE THREAD

**Issue by smiller171** [Yubico/pam-u2f#42]
I got PAM working on my Mac a few days ago, but today at work I notice it's randomly stopped working. Debug output below:

```
sh-3.2# su smiller
debug: pam-u2f.c:64 (parse_cfg): called.
debug: pam-u2f.c:65 (parse_cfg): flags 0 argc 1
debug: pam-u2f.c:67 (parse_cfg): argv[0]=debug
debug: pam-u2f.c:68 (parse_cfg): max_devices=0
debug: pam-u2f.c:69 (parse_cfg): debug=1
debug: pam-u2f.c:70 (parse_cfg): interactive=0
debug: pam-u2f.c:71 (parse_cfg): cue=0
debug: pam-u2f.c:72 (parse_cfg): manual=0
debug: pam-u2f.c:73 (parse_cfg): nouserok=0
debug: pam-u2f.c:74 (parse_cfg): alwaysok=0
debug: pam-u2f.c:75 (parse_cfg): authfile=(null)
debug: pam-u2f.c:76 (parse_cfg): origin=(null)
debug: pam-u2f.c:77 (parse_cfg): appid=(null)
debug: pam-u2f.c:119 (pam_sm_authenticate): Origin not specified, using "pam://Scotts-MBP-2"
debug: pam-u2f.c:130 (pam_sm_authenticate): Appid not specified, using the same value of origin (pam://Scotts-MBP-2)
debug: pam-u2f.c:140 (pam_sm_authenticate): Maximum devices number not set. Using default (24)
debug: pam-u2f.c:158 (pam_sm_authenticate): Requesting authentication for user smiller
debug: pam-u2f.c:169 (pam_sm_authenticate): Found user smiller
debug: pam-u2f.c:170 (pam_sm_authenticate): Home directory for smiller is /Users/smiller
debug: pam-u2f.c:177 (pam_sm_authenticate): Variable XDG_CONFIG_HOME is not set. Using default value ($HOME/.config/)
debug: pam-u2f.c:209 (pam_sm_authenticate): Using default authentication file /Users/smiller/.config/Yubico/u2f_keys
debug: util.c:107 (get_devices_from_authfile): Authorization line: smiller:AKyp1zw9XB7pCU03aKbSyNekjzX0Eeh1C7KY3v9eEXtR4poT7mHoioq8hp21ckzqxi4zF1Di7vXGSFw9zDlKVA,046f2691aaabc25e65e35130dbcaa42d532fa2ae1f9b6310a259143680d1376686dfb91506da83802325f82c15ec5f5799641c94f3cb63b621bcd68a33063eecb1
debug: util.c:112 (get_devices_from_authfile): Matched user: smiller
debug: util.c:130 (get_devices_from_authfile): KeyHandle for device number 1: AKyp1zw9XB7pCU03aKbSyNekjzX0Eeh1C7KY3v9eEXtR4poT7mHoioq8hp21ckzqxi4zF1Di7vXGSFw9zDlKVA
debug: util.c:157 (get_devices_from_authfile): publicKey for device number 1: 046f2691aaabc25e65e35130dbcaa42d532fa2ae1f9b6310a259143680d1376686dfb91506da83802325f82c15ec5f5799641c94f3cb63b621bcd68a33063eecb1
debug: util.c:172 (get_devices_from_authfile): Length of key number 1 is 65
debug: util.c:200 (get_devices_from_authfile): Found 1 device(s) for user smiller
debug: util.c:262 (do_authentication): Device max index is 0
debug: util.c:288 (do_authentication): Attempting authentication with device number 1
debug: util.c:310 (do_authentication): Challenge: { "keyHandle": "AKyp1zw9XB7pCU03aKbSyNekjzX0Eeh1C7KY3v9eEXtR4poT7mHoioq8hp21ckzqxi4zF1Di7vXGSFw9zDlKVA", "version": "U2F_V2", "challenge": "IEMNMAlXKY3DTyk_iru5InxI2s4Pi77rDGZg8SKVro0", "appId": "pam:\/\/Scotts-MBP-2" }
debug: util.c:328 (do_authentication): Unable to communicate to the device, authenticator error
debug: pam-u2f.c:256 (pam_sm_authenticate): do_authentication returned -2
debug: pam-u2f.c:275 (pam_sm_authenticate): done. [authentication error]
```

**Comment by a-dma** [comment]
I've tried reproducing this on a Mac, but to no avail.
Have you made any changes lately? I know there has been some work on libhiapi which libu2f-host uses. Could that be the culprit?

Maybe try rebuilding with libu2f-host in verbose mode.

**Comment by smiller171** [comment]
How do I go about doing that?

**Comment by a-dma** [comment]
We should probably have a better way, but for now you have to change the `0` to a `1` in `u2fh_global_init` here https://github.com/Yubico/pam-u2f/blob/master/util.c#L244

**Comment by smiller171** [comment]
@a-dma does that require a recompile or is there somewhere I can change it on the existing install? I just installed from Homebrew.

**Comment by a-dma** [comment]
Yes, unfortunately it does require you to recompile the module (I've opened #43 to address this), but so far that's the only solution I'm afraid.

**Comment by smiller171** [comment]
@a-dma is that not what this is? https://github.com/Yubico/pam-u2f/blob/master/pamu2fcfg/cmdline.c#L42

**Comment by a-dma** [comment]
Not really. That sets verbose mode in the registration tool (aka `pamu2fcfg`). Libu2f-host itself can't be set in debug mode without recompiling.

What you could try to do is use the tool that comes with libu2f-host (aka `u2f-host`) with the `-d` flag and try to emulate what the pam module does to see if you can recreate the problem.

**Comment by smiller171** [comment]
like this? `u2f-host -d --action=authenticate`

**Comment by smiller171** [comment]
oh, my libu2f-host and libu2f-server versions seem to be out of date they are on 1.0.0 and 1.0.1, respectively

**Comment by smiller171** [comment]
I modified the homebrew install script to get libu2f-host up to date. Still not working, but I'll try rebooting in a bit.

**Comment by a-dma** [comment]
yes you would want to do something like
`u2f-server -a register -ohttp://test.com -ihttp://test.com -k hndle.dat -p user.dat -d`
copy the challenge and paste it into
`u2f-host -ohttp://test.com -a register -d`

and then do a similar process with authentication

**Comment by elyscape** [comment]
This is indeed due to an outdated libu2f-host. I have submitted a pull request bumping Homebrew's version. Doing this should probably be added to the release process on Yubico's end.

**Comment by smiller171** [comment]
@elyscape it still didn't work when I updated libu2f-host.

The Apple Store replaced the logic board on my Mac for unrelated issues, and after re-installing all my stuff I had the same issue. Pam-u2f worked for a few days, then randomly stopped.

**Comment by elyscape** [comment]
Interesting. For me, it wasn't working at all until I updated libu2f-host, but now it works fine. I do need to hit the button for it to authenticate, though.

**Comment by smiller171** [comment]
@elyscape works for me for a couple days without updating libu2f-host, then stops randomly, and updating the lib doesn't help.

**Comment by g5pw** [comment]
I have the same issue, libu2f-{host,server} and pam-u2f installed from Macports.

**Comment by g5pw** [comment]
Yep, that's the issue. From @smiller171 's log:
> debug: pam-u2f.c:76 (parse_cfg): origin=(null)
> debug: pam-u2f.c:77 (parse_cfg): appid=(null)
> debug: pam-u2f.c:119 (pam_sm_authenticate): Origin not specified, using "pam://Scotts-MBP-2"
> debug: pam-u2f.c:130 (pam_sm_authenticate): Appid not specified, using the same value of origin (pam://Scotts-MBP-2)

The hostname has a `-2` suffix indicating the MBP probably renamed itself due to a 'false' .local name collision (it happens sometimes). This means it will probably fail again, next time it happens. IMO, the documentation should reflect the fact that `origin` and `appid` **have** to be set. Otherwise, if the hostname changes, authentication fails.

**Comment by smiller171** [comment]
Will need to find time to test, but this sounds promising @g5pw

**Comment by thorduri** [comment]
README has been updated, and this (DHCP setting a new hostname) seems to have been the issue.

**Comment by c00kiemon5ter** [comment]
Hello, I'm also using mac os x, and had the same problem. It is now that I solved it that I could  search the right keywords and see others having the same problem. For the sake of explanation and to make up for the searching I did, here is a description and the solution (tldr; @g5pw is right):

By default mac os x changes the hostname by doing a reverse dns lookup on the (dhcp) locally assigned ip. In order to reproduce the problem all that is needed is to setup your `~/.config/Yubicon/u2f_keys` when connected to some network, test that the key works and then disconnect (or connect to another network) - now the key should not work. 

The solution is to set a hostname, so that the OS does not change it when connecting or disconnecting between networks. The programmable way to do that, is by using `scutil` (system configuration utility) to query and set the hostname.

    sudo scutil -set HostName my-static-hostname-or-fqdn-that-will-not-change
    scutil -get HostName
    > my-static-hostname-or-fqdn-that-will-not-change

Cheers 🍺 🏖 

PS: one can always reset the default behaviour by specifying an empty hostname
